### PR TITLE
Move soft keyword definitions outside the parser

### PR DESCRIPTION
**Pull request by tgodzik**
Based on https://github.com/scalameta/scalafmt/pull/2284#discussion_r538886558

This way they can be reused inside scalafmt.

### CODE REVIEW

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import scala.meta.internal.classifiers.classifier
+
+class SoftKeywords(dialect: Dialect) {
+
+  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
+    case Ident(value) if pred(value.stripPrefix("`").stripSuffix("`")) => true
+    case _ => false
+  }
+
+  def isSoftKw(token: Token, name: String): Boolean = isIdentAnd(token, _ == name)
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:15]: Can this be `private`?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
+package scala.meta.internal.parsers
+
+import scala.meta.Dialect
+import scala.meta.tokens.Token
+import scala.meta.tokens.Token.Ident
+import scala.meta.internal.classifiers.classifier
+
+class SoftKeywords(dialect: Dialect) {
+
+  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:10]: Make private?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 3]
```diff
@@ -837,77 +840,6 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
...
-  }
-
-  @classifier
-  trait SkInfix {
-    val name = "infix"
-    def unapply(token: Token): Boolean = {
-      isSoftKw(token, name) && dialect.allowInfixMods
-    }
-  }
-
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:910]: why not extract all the other classifiers, too? lots of them are obviously useful to like whitespace, trivia etc.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    case Ident(value) if pred(value.stripPrefix("`").stripSuffix("`")) => true
+    case _ => false
+  }
+
+  def isSoftKw(token: Token, name: String): Boolean = isIdentAnd(token, _ == name)
+
+  @classifier
+  trait SkAs {
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:18]: Nitpick: I'm not a fan of the "Sk" prefix. Instead of using wildcard imports, how do you like this syntax?
```suggestion
class SoftKeywords {
  trait KwAs { ... }
}

val soft = new SoftKeywords()
x match {
  case soft.KwAs
```

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    val name = "as"
+    def unapply(token: Token): Boolean = {
+      isSoftKw(token, name) && dialect.allowAsPatternBinding
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:21]: Nit: can we move `dialect.allowAsPatternBinding` before the `isSoftKw`? The `isSoftKw` test is a lot more expensive and will be `false` in most cases

**Review by olafurpg — CHANGES_REQUESTED** [review]

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 5 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:21]: Sure! Fun fact, this was added with M1 and will actually be removed with M3.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+import scala.meta.Dialect
+import scala.meta.tokens.Token
+import scala.meta.tokens.Token.Ident
+import scala.meta.internal.classifiers.classifier
+
+class SoftKeywords(dialect: Dialect) {
+
+  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
+    case Ident(value) if pred(value.stripPrefix("`").stripSuffix("`")) => true
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:11]: very optional:
```
  case Ident(value) => pred(value.stripPrefix("`").stripSuffix("`"))
```

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 7]
```diff
@@ -837,77 +840,6 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
       (isCaseClassOrObject || (token.is[KwCase] && token.next.is[Ident] && dialect.allowEnums))
   }
 
-  @classifier
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:840]: (stupid github only allows commenting on red/green lines) should we remove `isSoftKw` above?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 2 above)
Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:10]: actually, this duplicates the method in `ScalametaParser` above. should we do this instead:

```
private[parsers] object ScalametaParser {
  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
    case Ident(value) => pred(value.stripPrefix("`").stripSuffix("`"))
    case _ => false
  }
}
```
and then import it in both classes?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 1 above)
Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:15]: also, could add dialect check as well, so it's always in the right order of checking:
```
  private matches(token: Token, name: String, isEnabled: Dialect => Boolean): Boolean =
    isEnabled(dialect) && isIdentAnd(token, _ == name)
```
(in this class, everything is about soft keywords, so i also renamed the method)

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala [hunk 8]
```diff
@@ -0,0 +1,87 @@
+package scala.meta.internal.parsers
+
+import scala.meta.Dialect
+import scala.meta.tokens.Token
+import scala.meta.tokens.Token.Ident
+import scala.meta.internal.classifiers.classifier
+
+class SoftKeywords(dialect: Dialect) {
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:8]: do we need to wrap this in an instance? wouldn't this work, with implicit:
```
  object SoftKw {
    @classifier
    trait Using {
      def unapply(token: Token)(implicit dialect: Dialect): Boolean = {
        dialect.allowGivenUsing && isSoftKw(token, "using")
      }
    }
  }

  // somewhere else
  token.is[SoftKw.Using]
```

**Review by kitbellew — CHANGES_REQUESTED** [review]
what started as a simple refactor...

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 8 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:8]: Tried it and it does not recognize it as an unapply method :/

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 3 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:910]: Most of it is internal to parser, so to be honest I am not sure what would make sense to extract. I will start with the soft keywords now, but let me know if you find anything that might be particularily useful in scalafmt.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 4 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:18]: Looks good, I will change it!

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 1 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala]: Moved it actually to ScalametParserObject and is now reused in SoftKeywords and ScalametaParser.

Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala]: Ach no, not that one :sweat_smile: this one is renamed to `matches` and made private.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 2 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala]: Moved it actually to ScalametParserObject and is now reused in SoftKeywords and ScalametaParser.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 4 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala]: Reafctored!

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 5 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala]: Used the advice from @kitbellew and now we have a matches method that first always checks if a particular dialect feature is supported.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala (hunk 6 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala]: Fixed, looks much nicer.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 7 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:840]: Ach right! Removed it!

**Comment by tgodzik** [comment]
> also, could add dialect check as well, so it's always in the right order of checking:
>
>  private matches(token: Token, name: String, isEnabled: Dialect => Boolean): Boolean =
>    isEnabled(dialect) && isIdentAnd(token, _ == name)
> (in this class, everything is about soft keywords, so i also renamed the method)

Fixed it, great suggestion! 

> actually, this duplicates the method in ScalametaParser above. should we do this instead:

Moved it, thanks! @kitbellew

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 9]
```diff
@@ -36,6 +36,8 @@ import scala.util.Failure
 class ScalametaParser(input: Input, dialect: Dialect) { parser =>
   require(Set("", EOL).contains(dialect.toplevelSeparator))
   implicit val currentDialect: Dialect = dialect
+  val soft = new SoftKeywords(currentDialect)
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:39]: somewhat optional: 1. `private` 2. use `dialect` directly, and not the intermediate `currentDialect`?

also, for a completely different, unrelated change: we could change the class to
```
class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
   ...
```
and remove `currentDialect`.

**Review by kitbellew — APPROVED** [review]

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 9 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Good idea! Changed!

**Review by kitbellew — APPROVED** [review]

**Comment by kitbellew** [comment]
@tgodzik should we release and use in `scalafmt`?

**Comment by tgodzik** [comment]
Sure, makes sense. We have two more PRs lined up, but it shouldn't really influence scalafmt. I will run it now.
